SMEE Aanderaa 5990 Conductivity App (Mote)

uma-kat opened the pull request.

towynlin commented on CMakePresets.json:
```diff
@@ -245,6 +245,13 @@
             }
+        },
+        {
+            "name": "aanderaa-conductivity-rs232",
```
Thank you for adding a preset! ❤️

towynlin commented on src/apps/rs232_expander_apps/aanderaa_conductivity/user_code/aanderaa_conductivity_sensor.h:
```diff
@@ -0,0 +1,69 @@
+    static constexpr ValueType PARSER_VALUE_TYPE[] = {TYPE_DOUBLE, TYPE_DOUBLE, TYPE_DOUBLE, TYPE_DOUBLE, TYPE_DOUBLE};
+    static constexpr char SENSOR_BM_LOG_ENABLE[] = "sensorBmLogEnable";
+    static constexpr char SENSOR_DEPTH_M[] = "sensorDepthM";
+    static constexpr char SENSOR_PRESSURE_KPA[] = "pressureKpa";
```
Let's remove this `pressureKpa` line, since we're exclusively sticking with the `sensorDepthM` config.

towynlin commented on src/apps/rs232_expander_apps/aanderaa_conductivity/user_code/aanderaa_conductivity_sensor.h:
```diff
@@ -0,0 +1,69 @@
+    static constexpr char SENSOR_BM_LOG_ENABLE[] = "sensorBmLogEnable";
+    static constexpr char SENSOR_DEPTH_M[] = "sensorDepthM";
+    static constexpr char SENSOR_PRESSURE_KPA[] = "pressureKpa";
+    static constexpr char SENSOR_INTERVAL_S[] = "readingIntervalS";
```
Per internal discussion, let's change to `readingPeriodS`.

towynlin commented on src/apps/rs232_expander_apps/aanderaa_conductivity/user_code/aanderaa_conductivity_sensor.h:
```diff
@@ -0,0 +1,69 @@
+    static constexpr char SENSOR_DEPTH_M[] = "sensorDepthM";
+    static constexpr char SENSOR_PRESSURE_KPA[] = "pressureKpa";
+    static constexpr char SENSOR_INTERVAL_S[] = "readingIntervalS";
+    static constexpr char SENSOR_CELL_COEFF[] = "cellCoeff";
```
Remove the final "f" so that we're consistent with the 5990 documentation: `cellCoef` with one "f".

towynlin commented on src/apps/rs232_expander_apps/aanderaa_conductivity/user_code/aanderaa_conductivity_sensor.cpp:
```diff
@@ -0,0 +1,222 @@
+  // reading interval in seconds
+  get_config_uint(BM_CFG_PARTITION_SYSTEM, SENSOR_INTERVAL_S, strlen(SENSOR_INTERVAL_S),
+                   &_intervalS);
+  printf("readingIntervalS: %" PRIu32 "\n", _intervalS);
```
Change all the words here (prints, variable names, constant names) to refer to "reading period".

towynlin commented on src/apps/rs232_expander_apps/aanderaa_conductivity/user_code/aanderaa_conductivity_sensor.cpp:
```diff
@@ -0,0 +1,222 @@
+  // calibration coefficients [PLACE HOLDER]
+  get_config_uint(BM_CFG_PARTITION_SYSTEM, SENSOR_CELL_COEFF, strlen(SENSOR_CELL_COEFF),
+                   &_cellCoeff);
+  printf("cellCoeff: %" PRIu32 "\n", _cellCoeff);
```
For consistency with the 5990 documentation, let's always abbreviate coefficient with a single F like CELL_COEF, cellCoef, etc.

victorsowa12 commented on src/apps/rs232_expander_apps/aanderaa_conductivity/user_code/aanderaa_conductivity_sensor.cpp:
```diff
@@ -0,0 +1,222 @@
+}
+
+void AanderaaConductivitySensor::configureSensor(void) {
+  //timeout is 60000 ms, looks like we gotta wake the senor and then send any commands after this.
```
I don't see a 60000ms delay any where. is this comment still relevant?

victorsowa12 commented on src/apps/rs232_expander_apps/aanderaa_conductivity/user_code/aanderaa_conductivity_sensor.cpp:
```diff
@@ -0,0 +1,222 @@
+  }
+  vTaskDelay(pdMS_TO_TICKS(2000));
+  // set cell coefficient command
+  */
```
is this commented out block still needed?

victorsowa12 commented on src/apps/rs232_expander_apps/aanderaa_conductivity/user_code/user_code.cpp:
```diff
@@ -0,0 +1,47 @@
+void setup(void) {
+  aanderaa_conductivity_sensor.init();
+  aanderaa_conductivity_topic_str_len = createAanderaaConductivityDataTopic();
+  vTaskDelay(pdMS_TO_TICKS(2000)); // delay for development
```
is this delay still needed?

matt001k commented on src/apps/rs232_expander_apps/aanderaa_conductivity/user_code/aanderaa_conductivity_sensor.cpp:
```diff
@@ -0,0 +1,222 @@
+
+  // passkey command
+  PLUART::write((uint8_t *)CMD_SET_PASSKEY_1, strlen(CMD_SET_PASSKEY_1));
+  vTaskDelay(pdMS_TO_TICKS(100));
```
I am curious, are all of these delays necessary? It seems like this would take the sensor a significant amount of time to initialize.
 Is there a way to poll if the Conductivity sensor's status to determine if it is ready before moving on to the next command?
If the delays are necessary, instead of using raw number, 500, 100, etc. Could you store these numbers as variables or macros?
ex:
```
constexpr uint32_t command_delay_ticks = pdMS_TO_TICKS(100);
```

uma-kat replied:
Good catch! Adding long delays was not well thought out by me. I'll check if all those commands need so much delay. My initial guess is some can go, some commands like Stop need a bit more time. I'll review these.

victorsowa12 reviewed (commented):
Functionality looks good to me! I noticed some comments/commented out code seemed no longer needed

matt001k commented on src/apps/rs232_expander_apps/aanderaa_conductivity/user_code/aanderaa_conductivity_sensor.cpp:
```diff
@@ -0,0 +1,222 @@
+  // send get_all command and cross check if they were saved.
+  PLUART::write((uint8_t *)CMD_GET_ALL, strlen(CMD_GET_ALL));
+  uint16_t line_count = 0;
+  while (line_count < 25) {
+    if (PLUART::lineAvailable()) {
+    read_len = PLUART::readLine(_payload_buffer, sizeof(_payload_buffer));
+    printf("%.*s\n", read_len, _payload_buffer);
+    line_count++;
+    }
+  }
```
Does this need a timeout? What happens if for some reason all of the lines do not make it through due to some UART corruption or misbehavior from the sensor?

uma-kat replied:
I could add a timeout instead. 👍

matt001k commented on src/apps/rs232_expander_apps/aanderaa_conductivity/user_code/user_code.cpp:
```diff
@@ -0,0 +1,47 @@
+}
+void loop(void) {
+  // Read and handle line from sensor
+  static AanderaaConductivityMsg::Data d;
```
I think this variable does not need to be static! Not doing so will place it on the task's stack which should have enough space 👍

uma-kat replied:
I noticed that other apps like turbidity, pme do and rbr all use `static`, so I just went with that. We can always change it later when we update the others too.

matt001k commented on src/apps/rs232_expander_apps/aanderaa_conductivity/user_code/aanderaa_conductivity_sensor.cpp:
```diff
@@ -0,0 +1,222 @@
+                 "tick: %" PRIu64 ", rtc: %s, line: %.*s\n", uptimeGetMs(), rtc_time_str,
+                 read_len, _payload_buffer);
+    }
+    spotter_log_console(0, "conductivity | tick: %" PRIu64 ", rtc: %s, line: %.*s", uptimeGetMs(),
+              rtc_time_str, read_len, _payload_buffer);
+    printf("conductivity | tick: %" PRIu64 ", rtc: %s, line: %.*s\n", uptimeGetMs(), rtc_time_str,
+           read_len, _payload_buffer);
```
Small optimization, the formatting string could be saved as a constant variable and used in each log statement:

```
    constexpr char *fmt = "conductivity | tick: %" PRIu64 ", rtc: %s, line: %.*s";

    spotter_log_console(0, fmt, uptimeGetMs(),
              rtc_time_str, read_len, _payload_buffer);
    printf(fmt, uptimeGetMs(), rtc_time_str,
           read_len, _payload_buffer);
```

 or store the items in a string prior to the log statements:

```
    // Might be smaller than 128... adjust as needed
    char str[128] = "";

    snprintf(str, sizeof(str), "conductivity | tick: %" PRIu64 ", rtc: %s, line: %.*s", 
                  uptimeGetMs(), rtc_time_str,
                  read_len, _payload_buffer);
 
    spotter_log_console(0, str);
    printf(str);
```
The benefit of doing this allows the string to change in the future and only one location will have to change rather than two, the second option being even better at this.

uma-kat replied:
I'm just going with what we've been doing in other apps like turbidity and RBR. Are we planning to switch things up soon with the optimization you're suggesting? If so, maybe we could tackle this part at the same time. If I change it now, it'll be the odd one out compared to everything else.

matt001k replied:
That is no problem just  a suggestion for future maintainability (from experience I have made the oops of changing one string and not the other creating a bug).

matt001k commented on src/apps/rs232_expander_apps/aanderaa_conductivity/user_code/aanderaa_conductivity_sensor.cpp:
```diff
@@ -0,0 +1,317 @@
+      clearPayloadBuffer();
+
+      // read cellCoef from the sensor
+	  uint32_t read_duration_ms = 1000;
+  	  uint32_t start_time = pdTICKS_TO_MS(xTaskGetTickCount());
+      PLUART::write((uint8_t *)CMD_GET_CELL_COEF, strlen(CMD_GET_CELL_COEF));
+	  while ((pdTICKS_TO_MS(xTaskGetTickCount()) - start_time) < read_duration_ms) {
+   		 if (PLUART::lineAvailable()) {
+      		read_len = PLUART::readLine(_payload_buffer, sizeof(_payload_buffer));
+      		if (read_len > 5) {
+        	  debug_printf("%.*s\n", read_len, _payload_buffer);
+      		  // cellCoef\t5990\t67\t4.585078E+00\r\n
+              // _cellCoef = 4.585078E+00
+      		  char *last_tab = strrchr(_payload_buffer, '\t');
+      		  if (last_tab != NULL) {
+      		    _cellCoef = strtof(last_tab + 1, NULL); // +1 to skip the tab
+      		    debug_printf("Measured cellCoef: %f\n", _cellCoef);
+      		  } else {
+      		    debug_printf("Failed to find tab separator\n");
+      		  }
+        	  clearPayloadBuffer();
+      	   }
+        }
+  	  }
```
Not necessary now, but in the future it would be nice to have a function called `read_data_from_sensor` that would capture a lot of this logic (as it is re-used in multiple locations).

uma-kat replied:
I'm on board with that idea! I could do that right away. I might ask you for a re-review after the push.

uma-kat replied:
Done!

matt001k replied:
Nice! Looks good!

matt001k reviewed (approved):
Nice work!

uma-kat commented:
@towynlin made some updates to the code after all the feedback.
- Added `_sensorBmLogEnable` at another place in the init function to log all the properties set on the sensor before sampling begins. 
- Added power bleed in `setup()` and set `ftl_recovery_ms` to 1000ms after chatting with Michael Bella. He said that's a safe value. `ftl_recovery_ms` is a configurable value same as it is defined in the RBR application
- Refactored `printf` to `debug_printf` appropriately

Could you please review this PR with these updates?

towynlin reviewed (approved):
Great work! Sorry — didn't realize my review was blocking. Ship it!
